kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/filesystem/FileSystemMetadataProvider.java

```diff
@@ -534,6 +535,16 @@ static void writeDescriptor(
         throw new DatasetIOException("Cannot close", e);
       }
     }
+
+    //rename the file into its final location so that changes appear atomic
+      try {
+          fs.rename(tempDescriptorPath, descriptorPath);
```

Comment:
Filesystem.rename() is a somewhat awkward API that has a boolean return value.  It doesn't work the way you might expect.  Interestingly the JavaDoc doesn't say anything about when it might return false but I don't think it will do the overwrite.  I believe what you need to do an atomic rename with overwrite is FileContext.rename(src, dest, Options.Rename.OVERWRITE).